### PR TITLE
Actually show deprecation warnings and their line level [2/2]

### PR DESCRIPTION
**Pull request by carmocca**
### Issue

DeprecationWarnings weren't being shown for two reasons:
  1. Need to set the stacklevel
  2. DeprecationWarnings needed to be enabled

### Solution

- Set the appropriate stacklevel for most calls
- Creates our own `LightningDeprecationWarning` and enable it
- Groups warning-related functions into `warnings.py`
- Add a stderr special test for warnings
 
## To decide

Since `DeprecationWarning`s haven't been shown, we need to decide if we should delay their removal versions.

Option 👀 : do nothing - an argument is that we already mention what's been deprecated in the release notes
Option 🚀 : Only delay the removal version of warnings set to be removed in 1.4
Option ❤️ : Do it for all

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 1]
```diff
@@ -113,30 +113,28 @@ def attach_dataloaders(
     def attach_datamodule(
         self, model: 'pl.LightningModule', datamodule: Optional['pl.LightningDataModule'] = None
     ) -> None:
-        # We use datamodule if it's been provided, otherwise we check model for it
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/data_connector.py:116]: unrelated changes?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -370,10 +370,10 @@ def extra(self, extra: Mapping[str, Any]) -> None:
 
         def check_fn(v):
             if v.grad_fn is not None:
-                warning_cache.warn(
+                warning_cache.deprecation(
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:373]: name suggestion "deprecate" in line with "warn"

File: pytorch_lightning/utilities/warnings.py [hunk 3]
```diff
@@ -11,17 +11,39 @@
...
+
+from pytorch_lightning.utilities.distributed import rank_zero_only
+
+
+def _warn(*args, stacklevel: int = 2, **kwargs):
+    warnings.warn(*args, stacklevel=stacklevel, **kwargs)
+
+
+@rank_zero_only
+def rank_zero_warn(*args, stacklevel: int = 4, **kwargs):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/warnings.py:25]: why are we splitting these apart?
can you keep them in the same file?

File: tests/utilities/test_warnings.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
```
Comment (awaelchli) [inline tests/utilities/test_warnings.py:1]: accidentally here?

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/data_connector.py:116]: `model.datamodule` is deprecated so I removed the deprecated usage otherwise it gets printed everywhere

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:373]: We already have `rank_zero_deprecation` instead of `rank_zero_deprecate`. Just following that

File: pytorch_lightning/utilities/warnings.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/warnings.py:25]: 1. Because it only makes sense that warnings-related stuff goes into the warnings.py files
2. This new structure reduces the circular conflicts (although there are still some).

If I keep the previous structure, then `distributed.py` would have needed to import `LightningDistributedWarning` but the `warning_cache` already imports `rank_zero_{warn,deprecation}` from distributed.py.

That is unless you think `LightningDistributedWarning` should be defined in `distributed.py`.

Now the only conflicts is that `warnings.py` needs `rank_zero_only` from `distributed.py` and `register_ddp_comm_hook` needs to warn.

IMO warnings are good where they are in this PR and the logging constructs should be moved out of `distributed.py`

File: tests/utilities/test_warnings.py (hunk 4 above)
Comment (carmocca) [reply tests/utilities/test_warnings.py:1]: wdym? This is the test for the changes in this PR

Comment (awaelchli) [reply tests/utilities/test_warnings.py:1]: what is the shell bang for if this is a python script?
if this is crucial to the fix or how tests are executed it would be good to learn why so we can remember in the future.

Comment (carmocca) [reply tests/utilities/test_warnings.py:1]: aah i'm blind

bad copy pasta

```suggestion
```

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/data_connector.py:116]: so you removed `datamodule = datamodule or getattr(model, 'datamodule', None)`
and pulled out `if datamodule`

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:373]: so warn -> warning then?

**Review by awaelchli — APPROVED** [review]
not fully convinced about splitting up the the rank_zero functions but if circular imports are trouble, ok sounds fair :)

File: tests/utilities/test_warnings.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    assert "test_warnings.py:31: UserWarning: test1" in output
+    assert "test_warnings.py:32: DeprecationWarning: test2" in output
+
+    assert "test_warnings.py:34: UserWarning: test3" in output
+    assert "test_warnings.py:35: DeprecationWarning: test4" in output
+
+    assert "test_warnings.py:37: LightningDeprecationWarning: test5" in output
+
+    assert "test_warnings.py:40: UserWarning: test6" in output
+    assert "test_warnings.py:41: LightningDeprecationWarning: test7" in output
```
Comment (carmocca) [inline tests/utilities/test_warnings.py:52]:
```suggestion
    assert "test_warnings.py:30: UserWarning: test1" in output
    assert "test_warnings.py:31: DeprecationWarning: test2" in output

    assert "test_warnings.py:33: UserWarning: test3" in output
    assert "test_warnings.py:34: DeprecationWarning: test4" in output

    assert "test_warnings.py:36: LightningDeprecationWarning: test5" in output

    assert "test_warnings.py:39: UserWarning: test6" in output
    assert "test_warnings.py:40: LightningDeprecationWarning: test7" in output
```

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/data_connector.py:116]: Exactly

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:373]: I don't mind which one we go with, but needs to be done in a separate PR as `warning_cache.deprecation` was added in 1/2

File: .azure-pipelines/ipu-tests.yml [hunk 6]
```diff
@@ -53,12 +53,9 @@ jobs:
         export GIT_TERMINAL_PROMPT=1
         python -c "fname = 'requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if 'fairscale' not in line] ; open(fname, 'w').writelines(lines)"
         python -c "fname = 'requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)"
-
         python ./requirements/adjust_versions.py requirements/extra.txt
         python ./requirements/adjust_versions.py requirements/examples.txt
-
-        pip install --requirement ./requirements/devel.txt --upgrade-strategy only-if-needed
-
+        pip install . --requirement requirements/devel.txt
```
Comment (carmocca) [inline .azure-pipelines/ipu-tests.yml:58]: Need to install the package on IPUs just as we do on GPUs. To avoid:

```python
Traceback (most recent call last):
  File "tests/utilities/test_warnings.py", line 23, in <module>
    from pytorch_lightning.utilities.warnings import _warn, rank_zero_deprecation, rank_zero_warn, WarningCache
ModuleNotFoundError: No module named 'pytorch_lightning'
```

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -171,7 +171,7 @@ def datamodule(self) -> Any:
         warning_cache.deprecation(
             "The `LightningModule.datamodule` property is deprecated in v1.3 and will be removed in v1.5."
             " Access the datamodule through using `self.trainer.datamodule` instead.",
-            stacklevel=5,
+            stacklevel=6,
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:174]: mind write a comment on what the 6 means or refer to some docs?

File: pytorch_lightning/utilities/distributed.py [hunk 8]
```diff
@@ -65,22 +65,26 @@ def _get_rank() -> int:
 rank_zero_only.rank = getattr(rank_zero_only, 'rank', _get_rank())
 
 
-def _warn(*args, **kwargs):
-    warnings.warn(*args, **kwargs)
-
-
-def _info(*args, **kwargs):
+def _info(*args, stacklevel: int = 2, **kwargs):
+    if python_version() >= "3.8.0":
```
Comment (Borda) [inline pytorch_lightning/utilities/distributed.py:69]: can we add docstring here or at least refer `stacklevel` meaning/usage?

File: pytorch_lightning/utilities/warnings.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/utilities/warnings.py:25]: mind add a comment about this reasoning to code bc quite sure the next refactor will try to get them together...

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:174]: It's from python

> The third optional keyword argument is stacklevel, which defaults to 1. If greater than 1, the corresponding number of stack frames are skipped when computing the line number and function name set in the LogRecord created for the logging event. This can be used in logging helpers so that the function name, filename and line number recorded are not the information for the helper function/method, but rather its caller.

https://docs.python.org/3/library/logging.html#logging.Logger.debug

File: pytorch_lightning/utilities/distributed.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/utilities/distributed.py:69]: It's part of the Python logging API, but there's not a direct link to reference

Comment (Borda) [reply pytorch_lightning/utilities/distributed.py:69]: I would say adding this link is good https://github.com/PyTorchLightning/pytorch-lightning/pull/8002#discussion_r655528895

Comment (carmocca) [reply pytorch_lightning/utilities/distributed.py:69]: That doesn't link to the stacklevel description exactly, and you wouldn't want to link to the debug docs also for `info` and `warn`

File: tests/utilities/test_warnings.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+Needs to be run outside of `pytest` as it captures all the warnings.
+"""
+import os
+from contextlib import redirect_stderr
+from io import StringIO
+
+from pytorch_lightning.utilities.warnings import _warn, rank_zero_deprecation, rank_zero_warn, WarningCache
+
+running_special = os.getenv("PL_RUNNING_SPECIAL_TESTS", "0") == "1"
```
Comment (tchaton) [inline tests/utilities/test_warnings.py:25]: Really neat ! Amazing finding.

**Review by tchaton — APPROVED** [review]
LGMT !
